src/bynder-js-sdk.js

```diff
@@ -571,4 +573,327 @@ export default class Bynder {
+     * @param {Number} file.length - The length of the file to be uploaded
+     * @param {Object} file.data={} - An object containing the assets' attributes
+     * @return {Promise} The information of the uploaded file, including IDs and all final file urls.
+     */
+    uploadFile(file) {
+        const { body, filename, data } = file;
+        const { brandId } = data;
+        const isBuffer = Buffer.isBuffer(body);
+        const isStream = typeof body.read === 'function';
+        const length = isBuffer ? body.length : file.length;
+
+        if (!this.validURL()) {
+            return rejectURL();
+        }
+        if (!brandId || brandId === '') {
+            return rejectValidation('upload', 'brandId');
+        }
+        if (!filename || filename === '') {
+            return rejectValidation('upload', 'filename');
+        }
+        if (!body || (!isStream && !isBuffer)) {
+            return rejectValidation('upload', 'body');
+        }
+        if (!length || typeof length !== 'number') {
+            return rejectValidation('upload', 'length');
+        }
+
+        const getClosestUploadEndpoint = this.getClosestUploadEndpoint.bind(this);
+        const initUpload = this.initUpload.bind(this);
+        const registerChunk = this.registerChunk.bind(this);
+        const finalizeUpload = this.finalizeUpload.bind(this);
+        const pollUploadStatus = this.pollUploadStatus.bind(this);
+        const saveAsset = this.saveAsset.bind(this);
+
+        return Promise.all([
+            getClosestUploadEndpoint(),
+            initUpload(filename)
+        ])
+        .then((res) => {
+            const [endpoint, init] = res;
+            const uploadPath = init.multipart_params.key;
+
+            // count number of chunks
+            const CHUNK_SIZE = 1024 * 1024 * 5;
+            const chunks = Math.ceil(length / CHUNK_SIZE);
+
+            const uploadChunkToS3 = (chunkData, chunk) => {
+                const form = new FormData();
+                const params = Object.assign(init.multipart_params, {
+                    name: `${util.basename(uploadPath)}/p${chunk}`,
+                    chunk,
+                    chunks,
+                    Filename: `${uploadPath}/p${chunk}`,
+                    key: `${uploadPath}/p${chunk}`
+                });
+                Object.keys(params).forEach((key) => {
+                    form.append(key, params[key]);
+                });
+                form.append('file', chunkData);
+                const headers = Object.assign(form.getHeaders(), {
+                    'content-length': form.getLengthSync()
+                });
+                return axios.post(endpoint, form, { headers });
+            };
+
+            // sequentially upload chunks to AWS, then register them
+            let chunk = 0;
+            if (isBuffer) {
+                // buffer input
+                return new Promise((resolve, reject) => {
```

Comment:
These promises here, since they do the same but in different situations, should also be extracted to an upper level where we can better reuse the code and decrease its complexity. Maybe it should be defined at the beginning where you "prepare" the function defining some `const` and etc.

Reply:
Good point. I've refactored this into a new class method called `uploadFileInChunks` (commit 89503d8574deedef8a64c961a2ef9bdb47359526). Can you comment on that?